dbt/adapters/athena/session.py

```diff
@@ -2,18 +2,22 @@
 import time
 from datetime import datetime, timedelta, timezone
 from functools import cached_property
+from hashlib import md5
 from typing import Any, Dict, List
 from uuid import UUID
-
 import boto3
 import boto3.session
+import json
 
 from dbt.adapters.athena.config import get_boto3_config
-from dbt.adapters.athena.constants import DEFAULT_THREAD_COUNT, LOGGER
+from dbt.adapters.athena.constants import DEFAULT_THREAD_COUNT, LOGGER, SESSION_IDLE_TIMEOUT_MIN
+from dbt.adapters.athena.relation import AthenaRelation
 from dbt.contracts.connection import Connection
+from dbt.events.functions import get_invocation_id
 from dbt.exceptions import DbtRuntimeError
 
-spark_session_locks: Dict[UUID, threading.Lock] = {}
+invocation_id = get_invocation_id()
+spark_session_list: Dict[UUID, str] = {}
```

Comment:
concern: The reason I originally had a `threading.Lock` for each session is so that multiple dbt threads don't try to allocate the same session and get deadlocked waiting for that session when other sessions are available. How have you handled this scenario?

Reply:
let me edit the description with the explanation to the change i made for session management

Reply:
Updated the description of this PR to include more details about threads, sessions and timeout

Reply:
I am not an expert with threading library but with limited knowledge all that I am doing is using the self.Lock to make changes to the dictionary of session list and while accessing them to see if a session already exist or not, i am taking a copy variable of the list of session dictionary items and checking based on the description which is a hash built on engine configuration and concatenated with dbt invocation id.
Also I using this description to pass to the description field when we create a session. This helps us effectively searching for dbt invocation or engine configuration in aws console.

In short, all that i am trying to implement here is:
we will maintain a dictionary of session id and description
and we check if a new model in the run has same description (i.e., engine config) or not 
and if there is we use it if not create new one. I dont think we need listing sessions everytime rather this dictionary should help us drive the available sessions. We are also popping out terminated sessions and waiting on idle sessions. 

Please let me know if this makes sense.

Reply:
That does make sense. The only scenario I was afraid of was that multiple dbt threads will try to allocate the same session but it looks like using the self.Lock does alleviate that scenario. 

I'm a little unsure what `spark_session_load: Dict[UUID, int]` does. Is that to queue models into sessions that have the same config? How is the maximum load determined?

Reply:
Yeah it is just meant to manage the number of queries in queue.
I have restricted it to 1 to make sure thread concurrency is maintained (if users don't care thread concurrency, may be we can make this a config and users can set this up to whatever maximum)

But essentially it is meant to make sure no more than 1 query is in queue per session.
for ex: lets say we have 6 models and we are going with 3 threads config
model1: _configA_
model2: _configA_
model3: configB
model4: configB
model5: configB
model6: configB

With the 3 thread config, we create three sessions for the first 3 models to ensure thread concurrency of 3 is maintained (although two of the first three are sharing the same config A  we still create three sessions to maintain thread concurrency limit of 3). Now let's say all three models are completed and now to take up the next three, we only have one session with configB to run the next 3 models. Meaning we would queue 3 models to the same session and reducing the concurrency from 3 to 1. So in order to tackle this issue, I am using this spark_session_load variable to restricting to hold only 1 query per session and if there is already 1 query in queue per session, we will not hold the next two in the same session queue rather we will create 2 other sessions and maintain concurrency of 3 defined by threads config.

In short this variable ensures thread concurrency by restricting a maximum of 1 query in queue per session.

Reply:
Very nice! 

> we will not hold the next two in the same session queue rather we will create 2 other sessions and maintain concurrency of 3 defined by threads config

Small clarification, would we terminate the sessions created with configA in this scenario? Else we would have created 5 sessions but only 3 should be in use at any time defined by the threads parameter. I'm afraid we are not controlling how many sessions are created if we go down that route. 

For example, if the threads=16 and models=64, I can envision a scenario where the first 16 use a unique model configs so the remaining models have to create a session for their execution. That is a completely fine scenario but the total sessions created could become 64 and many of them might be IDLE too long.

Reply:
Yeah actually that is the reason I reduced the default (20 min) idle timeout session set by AWS to a hardcoded 10 minute. 
So within 10 min there should be a new model with the same configuration available to keep it alive and if not it will be terminated.
So yes model 1 with config A will be terminated if it doesn't find a model in the next to run within 10 minutes.

Do you prefer exposing this session idle timeout as a config variable instead of us hard coding to 10 min?